plugins/gitlab/shared/GitLabUtils.ts

```diff
@@ -0,0 +1,308 @@
+
+export const MRSchema = z.object({
+  iid: z.number(),
+  title: z.string(),
+  description: z.string().nullish(),
+  web_url: z.string().url(),
+  state: z.string(),
+  draft: z.boolean(),
+  merged_at: z.string().datetime().nullish(),
+  created_at: z.string().datetime(),
+  author: AuthorSchema,
+  labels: z.array(z.string()),
+});
+
+export type MR = z.infer<typeof MRSchema>;
+
+export class GitLabUtils {
+  private static clientId = env.GITLAB_CLIENT_ID;
+  private static gitlabUrl = env.GITLAB_URL ?? "https://gitlab.com";
+  private static apiBaseUrl = `${this.gitlabUrl}/api/v4`;
+  private static supportedResources = [
+    UnfurlResourceType.Issue,
+    UnfurlResourceType.PR,
+  ];
+
+  public static oauthUrl = `${this.gitlabUrl}/oauth`;
+  public static clientSecret = env.GITLAB_CLIENT_SECRET;
+
+  public static get url() {
+    return integrationSettingsPath("gitlab");
+  }
+
+  /**
+   * Generates the error URL for GitLab authorization errors.
+   *
+   * @param error - The error message to include in the URL.
+   * @returns The URL to redirect to upon authorization error.
+   */
+  public static errorUrl(error: string): string {
+    return `${this.url}?error=${encodeURIComponent(error)}`;
+  }
+
+  /**
+   * Generates the callback URL for GitLab OAuth.
+   *
+   * @param baseUrl - The base URL of the application.
+   * @param params - Optional query parameters to include in the callback URL.
+   * @returns The full callback URL.
+   */
+  public static callbackUrl(
+    { baseUrl, params }: { baseUrl: string; params?: string } = {
+      baseUrl: env.URL,
+      params: undefined,
+    }
+  ): string {
+    const callbackPath = "/api/gitlab.callback";
+    return params
+      ? `${baseUrl}${callbackPath}?${params}`
+      : `${baseUrl}${callbackPath}`;
+  }
+
+  /**
+   * Generates the authorization URL for GitLab OAuth.
+   *
+   * @param state - A unique state string to prevent CSRF attacks.
+   * @returns The full URL to redirect the user to GitLab's OAuth authorization page.
+   */
+  public static authUrl(state: string): string {
+    const params = new URLSearchParams({
+      client_id: this.clientId,
+      redirect_uri: this.callbackUrl(),
+      response_type: "code",
+      state,
+      scope: "read_api read_user",
+    });
+
+    return `${this.oauthUrl}/authorize?${params.toString()}`;
+  }
+
+  /**
+   * Generates the installation request URL.
+   *
+   * @returns The URL for installation requests.
+   */
+  public static installRequestUrl(): string {
+    return `${this.url}?install_request=true`;
+  }
+
+  /**
+   * Makes an authenticated API request to GitLab.
+   *
+   * @param accessToken - The access token for authentication.
+   * @param endpoint - The API endpoint path.
+   * @param params - Additional fetch options.
+   * @param query - Query parameters to include in the request.
+   * @returns The response data from the GitLab API.
+   */
+  public static async apiRequest({
```

Comment:
So many fetch calls, we should just use https://github.com/jdalrymple/gitbeaker perhaps